node/overseer/overseer-gen/proc-macro/src/impl_builder.rs

```diff
@@ -193,103 +351,61 @@ pub(crate) fn impl_builder(info: &OverseerInfo) -> proc_macro2::TokenStream {
 			}
 		}
 
-		/// Convenience alias.
-		type SubsystemInitFn<T> = Box<dyn FnOnce(#handle) -> ::std::result::Result<T, #error_ty> >;
-
-		/// Initialization type to be used for a field of the overseer.
-		enum FieldInitMethod<T> {
-			/// Defer initialization to a point where the `handle` is available.
-			Fn(SubsystemInitFn<T>),
-			/// Directly initialize the subsystem with the given subsystem type `T`.
-			Value(T),
-			/// Subsystem field does not have a value just yet.
-			Uninitialized
-		}
-
-		impl<T> ::std::default::Default for FieldInitMethod<T> {
-			fn default() -> Self {
-				Self::Uninitialized
-			}
-		}
-
 		#[allow(missing_docs)]
-		pub struct #builder #builder_generics {
+		pub struct #builder <S, #( #subsystem_placeholder_ty, )* #( #baggage_placeholder_ty, )*> {
 			#(
-				#subsystem_name : FieldInitMethod< #builder_generic_ty >,
+				#subsystem_name: #subsystem_placeholder_ty,
 			)*
 			#(
-				#baggage_name : ::std::option::Option< #baggage_ty >,
+				#baggage_name: #baggage_placeholder_ty,
 			)*
 			spawner: ::std::option::Option< S >,
 		}
 
-		impl #builder_generics Default for #builder #builder_generics {
-			fn default() -> Self {
+		impl<#builder_generics> #builder<S, #( FieldMissing::<#field_type>, )*>
+			#builder_where_clause
+		{
+			fn new() -> Self {
 				// explicitly assure the required traits are implemented
 				fn trait_from_must_be_implemented<E>()
 				where
 					E: std::error::Error + Send + Sync + 'static + From<#support_crate ::OverseerError>
 				{}
 
 				trait_from_must_be_implemented::< #error_ty >();
-
 				Self {
-				#(
-					#subsystem_name: Default::default(),
-				)*
-				#(
-					#baggage_name: None,
-				)*
+					#(
+						#field_name: FieldMissing::<#field_type>::default(),
+					)*
 					spawner: None,
```

Comment:
nit: I think this should also become a proper `Init`, `Missing`?

Reply:
Do you mean without `Field` part? I was under impression that it could clash with some other types in the code.

Reply:
Yep